tests/storage/test_event_federation.py

```diff
@@ -1049,6 +1049,46 @@ def test_get_insertion_event_backward_extremities_in_room_attempted_event_retry_
             backfill_event_ids, ["insertion_eventB", "insertion_eventA"]
         )
 
+    def test_repeated_backfill_failures_are_retried_appropriately(self) -> None:
+        """
+        A test that reproduces #13929 (Postgres only).
+
+        We should be able to backoff repeatedly, even if the backoff formula would tell
+        us to wait for more seconds than can be expressed in a 32 bit signed int.
+        """
+        setup_info = self._setup_room_for_insertion_backfill_tests()
+        room_id = setup_info.room_id
+
+        # Pretend that we have tried and failed 10 times to backfill event A.
+        for _ in range(10):
+            self.get_success(
+                self.store.record_event_failed_pull_attempt(
+                    room_id, "insertion_eventA", "fake cause"
+                )
+            )
+
+        # If the backoff periods grow without limit:
+        # After the first failed attempt, we would have backed off for 1 << 1 = 2 hours.
+        # After the second failed attempt we would have backed off for 1 << 2 = 4 hours,
+        # so after the 10th failed attempt we should backoff for 1 << 10 == 1024 hours.
+        # Wait 1100 hours just so we have a nice round number.
+        self.reactor.advance(datetime.timedelta(hours=1100).total_seconds())
+
+        # 1024 hours in milliseconds is 1024 * 3600000, which exceeds the largest 32 bit
+        # signed integer. The bug we're reproducing is that this overflow causes an
+        # error in postgres preventing us from fetching a set of backwards extremities
+        # to retry fetching.
+        backfill_points = self.get_success(
+            self.store.get_insertion_event_backward_extremities_in_room(room_id)
+        )
+
+        # We should aim to fetch A (the backoff period has expired) and to fetch B
+        # (we haven't tried to fetch it yet).
+        backfill_event_ids = [backfill_point[0] for backfill_point in backfill_points]
+        self.assertListEqual(
+            backfill_event_ids, ["insertion_eventB", "insertion_eventA"]
+        )
```

Comment:
I usually find it clearer to just use `assertEqual` instead of the more specific versions:

```suggestion
        self.assertEqual(
            backfill_event_ids, ["insertion_eventB", "insertion_eventA"]
        )
```

Reply:
Seems like it gives the same nice error so works for me:

```
twisted.trial.unittest.FailTest: Lists differ: ['b6', 'b5', 'b4', '2', 'b3', 'b2', 'b1'] != ['asdf', 'b6', 'b5', 'b4', '2', 'b3', 'b2', 'b1']

First differing element 0:
'b6'
'asdf'

Second list contains 1 additional elements.
First extra element 7:
'b1'

- ['b6', 'b5', 'b4', '2', 'b3', 'b2', 'b1']
+ ['asdf', 'b6', 'b5', 'b4', '2', 'b3', 'b2', 'b1']
?  ++++++++
```

Reply:
[`assertEqual` calls `assertListEqual` underneath](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertListEqual):

> These methods are used by default when comparing lists or tuples with `assertEqual()`.

Reply:
4f6637dbd